tutorials/mnist_blackbox.py

```diff
@@ -0,0 +1,250 @@
+flags.DEFINE_float('lmbda', 0.2, 'Lambda in https://arxiv.org/abs/1602.02697')
+
+
+def setup_tutorial():
+    """
+    Helper function to check correct configuration of tf and keras for tutorial
+    :return: True if setup checks completed
+    """
+
+    # Set TF random seed to improve reproducibility
+    tf.set_random_seed(1234)
+
+    if not hasattr(backend, "tf"):
+        raise RuntimeError("This tutorial requires keras to be configured"
+                           " to use the TensorFlow backend.")
+
+    # Image dimensions ordering should follow the Theano convention
+    if keras.backend.image_dim_ordering() != 'tf':
+        keras.backend.set_image_dim_ordering('tf')
+        print("INFO: '~/.keras/keras.json' sets 'image_dim_ordering' "
+              "to 'th', temporarily setting to 'tf'")
+
+    return True
+
+
+def prep_bbox(sess, x, y, X_train, Y_train, X_test, Y_test):
+    """
+    Define and train a model that simulates the "remote"
+    black-box oracle described in the original paper.
+    :param sess: the TF session
+    :param x: the input placeholder for MNIST
+    :param y: the ouput placeholder for MNIST
+    :param X_train: the training data for the oracle
+    :param Y_train: the training labels for the oracle
+    :param X_test: the testing data for the oracle
+    :param Y_test: the testing labels for the oracle
+    :return:
+    """
+
+    # Define TF model graph (for the black-box model)
+    model = model_mnist()
+    predictions = model(x)
+    print("Defined TensorFlow model graph.")
+
+    # Train an MNIST model
+    model_train(sess, x, y, predictions, X_train, Y_train, verbose=False)
+
+    # Print out the accuracy on legitimate data
+    accuracy = model_eval(sess, x, y, predictions, X_test, Y_test)
+    print('Test accuracy of black-box on legitimate test '
+          'examples: ' + str(accuracy))
+
+    return predictions
+
+
+def substitute_model(img_rows=28, img_cols=28, nb_classes=10):
+    """
+    Defines the model architecture to be used by the substitute
+    :param img_rows: number of rows in input
+    :param img_cols: number of columns in input
+    :param nb_classes: number of classes in output
+    :return: keras model
+    """
+    model = Sequential()
+
+    # Find out the input shape ordering
+    if keras.backend.image_dim_ordering() == 'th':
+        input_shape = (1, img_rows, img_cols)
+    else:
+        input_shape = (img_rows, img_cols, 1)
+
+    # Define a fully connected model (it's different than the black-box)
+    layers = [Flatten(input_shape=input_shape),
+              Dense(200),
+              Activation('relu'),
+              Dropout(0.5),
+              Dense(200),
+              Activation('relu'),
+              Dropout(0.5),
+              Dense(nb_classes),
+              Activation('softmax')]
+
+    for layer in layers:
+        model.add(layer)
+
+    return model
+
+
+def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads):
```

Comment:
This might be worth moving into the main cleverhans module sometime, but no need to hold up this PR